### DISCUSSION TITLE
Using GeoTiffs as elevation source

### DISCUSSION THREAD

**Discussion by Sgambe33** [iTowns/itowns#2275] [Q&A]
Hello, I was wondering if it is possible to use a GeoTiff as an elevation source. Obviously not directly but with maybe a GIS server. I tried using Geoserver with WMS service but the results were not good:
![image](https://github.com/iTowns/itowns/assets/80386324/6a13899f-9be2-42c6-82f3-32e080b04fc6)
As you can see there are a lot of empty spaces and inconsistencies between the different tiles. Is there anything else I can try?

**Comment by ketourneau** [comment]
It seems to be normal to see spaces between different tiles in certain situations (look at [3D mesh gap video](https://youtu.be/sXFxfqGPR-c?t=85)).
iTowns got parameter 'disableSkirt' on a Layer. By default the parameter is false so you normally will not see empty spaces.
Can you share your html ?

**Comment by vladimir-rybalko** [comment]
@Sgambe33, I can assume that you are using WMS with tiles mode. That's why Geoserver only applies your style within one tile and it doesn't know anything about the height of the adjacent tile. Try using [singleTile](https://docs.geoserver.org/main/en/user/services/wms/vendor.html#tiled) mode.

**Reply by Sgambe33** [reply]
@vladimir-rybalko How do I do it?

**Comment by Sgambe33** [comment]
Ok, I fixed the elevation problem but the map is still buggy. There are repeated tiles in wrong positions:
![image](https://github.com/iTowns/itowns/assets/80386324/0a1b5665-cf20-4d89-b69e-23db06268ef9)

**Reply by vladimir-rybalko** [reply]
@Sgambe33, can you share your html?

**Comment by Sgambe33** [comment]
@vladimir-rybalko Here:
```html
<body>
<div id="viewerDiv" class="viewer"></div>
<script>
    var extent = new itowns.Extent('EPSG:3857', 1119229.1867, 1133637.5665, 5474878.5880,5489822.0270);

    var viewerDiv = document.getElementById('viewerDiv');

    // Instanciate PlanarView
    // By default itowns' tiles geometry have a "skirt" (ie they have a height),
    // but in case of orthographic we don't need this feature, so disable it
    var view = new itowns.PlanarView(viewerDiv, extent, {
        disableSkirt: false, maxSubdivisionLevel: 18,
        placement: new itowns.Extent('EPSG:3857', 1119229.1867, 1133637.5665, 5474878.5880,5489822.0270),
        controls: {
            // Faster zoom in/out speed
            zoomFactor: 3,
            // prevent from zooming in too much
            maxResolution: 0.005  // a pixel shall not represent a metric size smaller than 5 mm
        },
    });

    // Add a TMS imagery source
    var opensmSource = new itowns.TMSSource({
        isInverted: false,
        // eslint-disable-next-line no-template-curly-in-string
        url: 'https://tile.openstreetmap.de/${z}/${x}/${y}.png',
        networkOptions: {crossOrigin: 'anonymous'},
        extent,
        crs: 'EPSG:3857',
        attribution: {
            name: 'OpenStreetMap',
            url: 'http://www.openstreetmap.org/',
        },
    });

    // Add a TMS imagery layer
    var opensmLayer = new itowns.ColorLayer('OPENSM', {
        updateStrategy: {
            type: itowns.STRATEGY_DICHOTOMY,
        },
        source: opensmSource,
    });

    view.addLayer(opensmLayer);
    
    // Request redraw

    const wmsSource = new itowns.WMSSource({
        url: 'http://localhost:8080/geoserver/ows?',
        version: '1.3.0',
        name: 'italia:dsm_4326',
        crs: 'EPSG:3857',
        transparent: true,
        width: 256,
        format: 'image/png',
        extent: extent,
        vendorSpecific: {
            'tiled': false
        }
    });
    const elevation_layer = new itowns.ElevationLayer('elevation', {
        source: wmsSource,
        useColorTextureElevation: true,
        colorTextureElevationMinZ: 203,
        colorTextureElevationMaxZ: 1187,

    });

    view.addLayer(elevation_layer);
    view.notifyChange();

</script>
</body>
```

**Reply by vladimir-rybalko** [reply]
Try using **WMS** explicitly in the service description http://localhost:8080/geoserver/wms
The elevation [layer](https://www.itowns-project.org/itowns/docs/#api/Layer/ElevationLayer) projection should be in EPSG:4326 = WGS84, as I remember. Change **CRS**.
```js
const wmsSource = new itowns.WMSSource({
        url: 'http://localhost:8080/geoserver/wms?',
        version: '1.3.0',
        service: 'WMS',
        name: 'italia:dsm_4326',
        crs: 'EPSG:4326',
        transparent: true,
        width: 256,
        format: 'image/png',
        vendorSpecific: {
            'tiled': false
        }
    });
```
Please keep us updated on changes.

**Reply by Sgambe33** [reply]
@vladimir-rybalko  Nothing changes
![image](https://github.com/iTowns/itowns/assets/80386324/b9e0ff43-f1ac-46aa-ae66-46d85c29b3dc)

**Reply by Desplandis** [reply]
@Sgambe33 Since you are using `geoserver`, it depends how your original elevation data (a single band geotiff I presume) is encoded as **png** which is the format distributed through WMS. Its been a long time since I last used `geoserver` but I recall you can configure it through [Styles](https://docs.geoserver.org/main/en/user/styling/webadmin/index.html). However, this suppose that both server and client (`itowns`) agree on the png encoding of the elevation data...

iTowns does not currently support `geotiff` as elevation source but this on the roadmap of the projet!
